### ISSUE TITLE
CCP Process Failed - Unable to get connection from pool

### ISSUE THREAD

**Issue by msweier** [opendcs/opendcs#480]
**Describe the bug**
A process failed in OpenDCS 7.0.11 from not being able to get a connection from pool (see log snippet below). I've never had a process fail out of the blue before and these comps/process has been running without issue for months/years. It was on the weekend so it failed completely on its own. I restarted the process, and it seems to be running along fine again.

```DBG2    02/10/24 20:28:11 Dev Msg: Plain create statement from:
DBG1    02/10/24 20:28:11 CwmsTimeSeriesDAO After fill, cache has 11450 TSIDs.
DBG1    02/10/24 20:28:12 CwmsTimeSeriesDAO reloadTsIdCache()
FAILURE 02/10/24 20:28:12 CwmsTimeSeriesDb Unable to get connection from pool: ORA-20998: ERROR: Unable to set a default SESSION_OFFICE_ID. The user:  either has no assigned CWMS database privileges or has privileges to more than one office. Please see your CWMS Application Admin.
ORA-06512: at "CWMS_20.CWMS_ERR", line 59
ORA-06512: at "CWMS_20.CWMS_ENV", line 49
ORA-28112: failed to execute policy function
ORA-06512: at "CWMS_20.CWMS_UTIL", line 626
ORA-06512: at "CWMS_20.CWMS_ENV", line 45
ORA-06512: at line 1

WARNING 02/10/24 20:28:12 GateCalcs_09-10 Unexpected exception while Getting new data: java.lang.RuntimeException: Error retrieving connection.
DBG1    02/10/24 20:28:12 runApp() exiting.
INFO    02/10/24 20:28:12 Closing database connection.
INFO    02/10/24 20:28:12 Shutting down Decodes Connection.
INFO    02/10/24 20:28:12 Closing  database connection.
INFO    02/10/24 20:28:12 GateCalcs_09-10 exiting.```

**Comment by MikeNeilson** [comment]
That's... odd. When I read the title I was expecting something different.

That particular error though implies something went wrong *after* the system actually got a connection. So you had a database hiccup and then OpenDCS decided it couldn't recover. Pretty sure we could from that one so may be able to handle the issue.

**Comment by msweier** [comment]
What's weird is only that process failed. Other processes have identical comps. There wasn't any data coming in other than the normal GOES stuff, so not sure why it would trip up.

**Comment by MikeNeilson** [comment]
I'm especially confused as the error `failed to execute policy function` generally means there was a package error.

**Comment by msweier** [comment]
This just happened again in 7.0.12. Not sure why.

`DBG1    05/29/24 16:55:42 CwmsTimeSeriesDAO reloadTsIdCache()
FAILURE 05/29/24 16:55:42 CwmsTimeSeriesDb Unable to get connection from pool: ORA-20998: ERROR: Unable to set a default SESSION_OFFICE_ID. The user:  either has no assigned CWMS database privileges or has privileges to more than one office. Please see your CWMS Application Admin.
ORA-06512: at "CWMS_20.CWMS_ERR", line 59
ORA-06512: at "CWMS_20.CWMS_ENV", line 63
ORA-02399: exceeded maximum connect time, you are being logged off
ORA-06512: at "CWMS_20.CWMS_UTIL", line 626
ORA-06512: at "CWMS_20.CWMS_ENV", line 59
ORA-06512: at line 1

WARNING 05/29/24 16:55:42 MissNavSum1 Unexpected exception while Getting new data: java.lang.RuntimeException: Error retrieving connection.
WARNING 05/29/24 16:55:42 MissNavSum1 java.lang.RuntimeException: Error retrieving connection.
	at decodes.cwms.CwmsTimeSeriesDb.getConnection(CwmsTimeSeriesDb.java:1650)
	at decodes.cwms.CwmsTimeSeriesDAO.getConnection(CwmsTimeSeriesDAO.java:1541)
	at opendcs.dao.DaoBase.doQuery(DaoBase.java:201)
	at decodes.cwms.CwmsTimeSeriesDAO.reloadTsIdCache(CwmsTimeSeriesDAO.java:1122)
	at decodes.cwms.CwmsTimeSeriesDAO.getNewData(CwmsTimeSeriesDAO.java:1280)
	at decodes.tsdb.ComputationApp.runApp(ComputationApp.java:418)
	at decodes.tsdb.TsdbAppTemplate.execute(TsdbAppTemplate.java:256)
	at decodes.tsdb.ComputationApp.main(ComputationApp.java:681)
Caused by: java.sql.SQLException: ORA-20998: ERROR: Unable to set a default SESSION_OFFICE_ID. The user:  either has no assigned CWMS database privileges or has privileges to more than one office. Please see your CWMS Application Admin.
ORA-06512: at "CWMS_20.CWMS_ERR", line 59
ORA-06512: at "CWMS_20.CWMS_ENV", line 63
ORA-02399: exceeded maximum connect time, you are being logged off
ORA-06512: at "CWMS_20.CWMS_UTIL", line 626
ORA-06512: at "CWMS_20.CWMS_ENV", line 59
ORA-06512: at line 1

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:509)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:461)
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1104)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:550)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:268)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:655)
	at oracle.jdbc.driver.T4CCallableStatement.doOall8(T4CCallableStatement.java:265)
	at oracle.jdbc.driver.T4CCallableStatement.doOall8(T4CCallableStatement.java:86)
	at oracle.jdbc.driver.T4CCallableStatement.executeForRows(T4CCallableStatement.java:965)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1205)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3666)
	at oracle.jdbc.driver.T4CCallableStatement.executeInternal(T4CCallableStatement.java:1358)
	at oracle.jdbc.driver.OraclePreparedStatement.execute(OraclePreparedStatement.java:3778)
	at oracle.jdbc.driver.OracleCallableStatement.execute(OracleCallableStatement.java:4251)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.execute(OraclePreparedStatementWrapper.java:1081)
	at oracle.ucp.jdbc.proxy.oracle$1ucp$1jdbc$1proxy$1oracle$1StatementProxy$2oracle$1jdbc$1internal$1OracleCallableStatement$$$Proxy.execute(Unknown Source)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:3416)
	at org.jooq.impl.AbstractRoutine.execute0(AbstractRoutine.java:521)
	at org.jooq.impl.AbstractRoutine.executeCallableStatement(AbstractRoutine.java:481)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:344)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:321)
	at usace.cwms.db.jooq.codegen.packages.CWMS_ENV_PACKAGE.call_SET_SESSION_OFFICE_ID(CWMS_ENV_PACKAGE.java:57)
	at usace.cwms.db.jooq.util.ConnectionPoolUcp.setSessionVariables_aroundBody58(ConnectionPoolUcp.java:521)
	at usace.cwms.db.jooq.util.ConnectionPoolUcp$AjcClosure59.run(ConnectionPoolUcp.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at usace.cwms.db.aspect.jooq.SQLExceptionWrapperAspect.translateToSQLException(SQLExceptionWrapperAspect.java:51)
	at usace.cwms.db.jooq.util.ConnectionPoolUcp.setSessionVariables(ConnectionPoolUcp.java:510)
	at usace.cwms.db.jooq.util.ConnectionPoolUcp.getConnection_aroundBody28(ConnectionPoolUcp.java:312)
	at usace.cwms.db.jooq.util.ConnectionPoolUcp$AjcClosure29.run(ConnectionPoolUcp.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:167)
	at usace.cwms.db.aspect.jooq.SQLExceptionWrapperAspect.translateToSQLException(SQLExceptionWrapperAspect.java:51)
	at usace.cwms.db.jooq.util.ConnectionPoolUcp.getConnection(ConnectionPoolUcp.java:297)
	at usace.cwms.db.dao.util.connection.CwmsDbConnectionPool.getConnection(CwmsDbConnectionPool.java:251)
	at usace.cwms.db.dao.util.connection.CwmsDbConnectionPool.getConnection(CwmsDbConnectionPool.java:277)
	at decodes.cwms.CwmsConnectionPool.getConnection(CwmsConnectionPool.java:250)
	at decodes.cwms.CwmsTimeSeriesDb.getConnection(CwmsTimeSeriesDb.java:1645)
	... 7 more
Caused by: Error : 20998, Position : 0, Sql = BEGIN "CWMS_20"."CWMS_ENV"."SET_SESSION_OFFICE_ID" (:1 ) ; END;, OriginalSql = { call "CWMS_20"."CWMS_ENV"."SET_SESSION_OFFICE_ID" (?) }, Error Msg = ORA-20998: ERROR: Unable to set a default SESSION_OFFICE_ID. The user:  either has no assigned CWMS database privileges or has privileges to more than one office. Please see your CWMS Application Admin.
ORA-06512: at "CWMS_20.CWMS_ERR", line 59
ORA-06512: at "CWMS_20.CWMS_ENV", line 63
ORA-02399: exceeded maximum connect time, you are being logged off
ORA-06512: at "CWMS_20.CWMS_UTIL", line 626
ORA-06512: at "CWMS_20.CWMS_ENV", line 59
ORA-06512: at line 1

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:513)
	... 43 more

DBG1    05/29/24 16:55:42 runApp() exiting.
INFO    05/29/24 16:55:42 Closing database connection.
INFO    05/29/24 16:55:42 Shutting down Decodes Connection.
INFO    05/29/24 16:55:42 Closing  database connection.
INFO    05/29/24 16:55:42 MissNavSum1 exiting.`

**Comment by MikeNeilson** [comment]
Hmm appears to be linking a connection somewhere. Was the nohup output redirected to a file? and if not can you set that up and let it run to failure again? When it exits normally like it appears to have, it'll print the location of any open connections that should've been closed.

**Comment by MikeNeilson** [comment]
Although, with that particular error it's more like it the system never actually closed the connection correctly.

**Comment by msweier** [comment]
> Hmm appears to be linking a connection somewhere. Was the nohup output redirected to a file? and if not can you set that up and let it run to failure again

No it was a different process this time. I can try and get them outputting to a log for all processes. I launched/interacted the client side GUI around that time. Although the first time we had this failure it was during the weekend with no one working so it's probably unrelated.

**Comment by MikeNeilson** [comment]
Do you have any computations, python or java, that use a JDBC connection to get data from the database. E.g. that calls `getConnection()`?

**Comment by msweier** [comment]
> Do you have any computations, python or java, that use a JDBC connection to get data from the database. E.g. that calls `getConnection()`?

Would [cwmspy](https://github.com/jetilton/cwmspy/blob/master/cwmspy/core.py) fit that bill? If so, then yes I do. Runs a few times an hour.

**Comment by MikeNeilson** [comment]
No, it would have to be an OpenDCS algorithm, being used by comproc, to have an impact. I should actually be able to recreate that locally faster *if* the issue is the connection not getting closed correctly, thus leading to the time exceeded error.

**Comment by msweier** [comment]
We have a fair amount of python algos and some custom java but I don't think any of them are using JDBC connection in the code.  Unless it's implicit in one of the stock python CCP Built In.

**Comment by MikeNeilson** [comment]
They shouldn't... but I should give that a look just in case.

**Comment by MikeNeilson** [comment]
Okay, the only built-in java algorithms that even reference the connection are some of the HDB connections, and PythonAlgorithm. If you have time to review your python algorithms I would do so, you're probably right that they don't.

**Comment by MikeNeilson** [comment]
Oh FYI, I lied about printing out the location of the connection. it will do that but you have to set an extra property before compproc runs.

**Comment by msweier** [comment]
I'll check the python rating() code. That one is used a lot.

**Comment by MikeNeilson** [comment]
Hmm, that could be it. I just looked at some of the code it and really doesn't handle connections well. I'm updating it now and it should end in the 7.0.13-RC03 release when that is put together.

**Comment by ktarbet** [comment]
Since reloadTsIdCache() is part of the stack trace, could some concurrency issue be at play, related to the cache?

**Comment by MikeNeilson** [comment]
I don't think so comproc operations are all on a single thread. Even the rating lookup at least appeared to be handling connections correctly but I still went and cleaned it up... along with the general group operations and computations so now I'm fixing more than I expected in the PR, but I'm pretty sure it's all related. I noticed a few original style doQueries calls that never got deleted even though the section was changed to try-with and prepared statements.
